sql/kv_test.go

```diff
@@ -0,0 +1,459 @@
+		kvs, pErr := txn.Scan(fmt.Sprintf("%s%06d", kv.prefix, 0), fmt.Sprintf("%s%06d", kv.prefix, rows), int64(rows))
+		if len(kvs) != rows {
+			return roachpb.NewErrorf("expected %d rows; got %d", rows, len(kvs))
+		}
+		return pErr
+	})
+	return pErr.GoError()
+}
+
+func (kv *kvNative) prep(rows int, initData bool) error {
+	kv.epoch++
+	kv.prefix = fmt.Sprintf("%d/", kv.epoch)
+	if !initData {
+		return nil
+	}
+	pErr := kv.db.Txn(func(txn *client.Txn) *roachpb.Error {
+		b := txn.NewBatch()
+		for i := 0; i < rows; i++ {
+			b.Put(fmt.Sprintf("%s%06d", kv.prefix, i), i)
+		}
+		return txn.Run(b)
+	})
+	return pErr.GoError()
+}
+
+func (kv *kvNative) done() {
+	kv.doneFn()
+}
+
+// kvSQL is a SQL-based implementation of the KV interface.
+type kvSQL struct {
+	db     *sql.DB
+	doneFn func()
+}
+
+func newKVSQL(b *testing.B) kvInterface {
+	enableTracing := tracing.Disable()
+	s := server.StartTestServer(b)
+	pgURL, cleanupURL := sqlutils.PGUrl(b, s, security.RootUser, "benchmarkCockroach")
+	pgURL.Path = "bench"
+	db, err := sql.Open("postgres", pgURL.String())
+	if err != nil {
+		b.Fatal(err)
+	}
+	if _, err := db.Exec(`CREATE DATABASE IF NOT EXISTS bench`); err != nil {
+		b.Fatal(err)
+	}
+
+	kv := &kvSQL{}
+	kv.db = db
+	kv.doneFn = func() {
+		db.Close()
+		cleanupURL()
+		s.Stop()
+		enableTracing()
+	}
+	return kv
+}
+
+func (kv *kvSQL) insert(rows, run int) error {
+	firstRow := rows * run
+	var buf bytes.Buffer
+	buf.WriteString(`INSERT INTO bench.kv VALUES `)
+	for i := 0; i < rows; i++ {
+		if i > 0 {
+			buf.WriteString(", ")
+		}
+		fmt.Fprintf(&buf, "('%06d', %d)", i+firstRow, i)
+	}
+	_, err := kv.db.Exec(buf.String())
+	return err
+}
+
+func (kv *kvSQL) update(rows, run int) error {
+	var buf bytes.Buffer
+	buf.WriteString(`UPDATE bench.kv SET v = v + 1 WHERE k IN (`)
+	for j := 0; j < rows; j++ {
+		if j > 0 {
+			buf.WriteString(", ")
+		}
+		fmt.Fprintf(&buf, `'%06d'`, j)
+	}
+	buf.WriteString(`)`)
+	_, err := kv.db.Exec(buf.String())
+	return err
+}
+
+func (kv *kvSQL) del(rows, run int) error {
+	var buf bytes.Buffer
+	buf.WriteString(`DELETE FROM bench.kv WHERE k IN (`)
+	for j := 0; j < rows; j++ {
+		if j > 0 {
+			buf.WriteString(", ")
+		}
+		fmt.Fprintf(&buf, `'%06d'`, j)
+	}
+	buf.WriteString(`)`)
+	_, err := kv.db.Exec(buf.String())
+	return err
+}
+
+func (kv *kvSQL) scan(count, run int) error {
+	rows, err := kv.db.Query(fmt.Sprintf("SELECT * FROM bench.kv LIMIT %d", count))
+	if err != nil {
+		return err
+	}
+	n := 0
+	for rows.Next() {
+		n++
+	}
+	rows.Close()
+	if err := rows.Err(); err != nil {
+		return err
+	}
+	if n != count {
+		return util.Errorf("unexpected result count: %d (expected %d)", n, count)
+	}
+	return nil
+}
+
+func (kv *kvSQL) prep(rows int, initData bool) error {
+	if _, err := kv.db.Exec(`DROP TABLE IF EXISTS bench.kv`); err != nil {
+		return err
+	}
+	if _, err := kv.db.Exec(`CREATE TABLE IF NOT EXISTS bench.kv (k STRING PRIMARY KEY, v INT)`); err != nil {
```

Comment:
Of course, this will prevent updates from working since we currently disallow changing the value of a primary key column. That limitation could be removed, though.